### ISSUE TITLE
[NEW] WAIT ALL

### ISSUE THREAD

**Issue by zuiderkwast** [madolson/placeholderkv#743]
I want to revive the feature suggested in https://github.com/redis/redis/pull/8684.

WAIT ALL in a cluster would provide a real consistency. Without waiting for all replicas, there's a risk that the one replica that didn't replicate all writes will win the failover election.

A problem with mentioned PR is that `WAIT ALL` would succeed if all replicas are gone, e.g. during a netsplit. In a cluster, we can avoid this problem by looking at the cluster structures rather than `server.replicas`. A replica that's gone missing is still present in the cluster until admin calls CLUSTER FORGET.

Can we do something similar for Sentinel? I'm not familiar enough with sentinel. Perhaps the sentinel knows and can tell the primary how many replicas it's supposed to have(?).

I care less about standalone primary-replica setups since they don't have the same guarantees anyway. I'd be fine with just counting `server.replicas` in this case (though some delayed decrement of the replica count would mitigate this a bit, perhaps). A manual config for the expected number of replicas can also be acceptable.

**Comment by enjoy-binbin** [comment]
Agree with this, i want to do that. Another thing i want to deal with is, if we pass a large numreplicas, can we implicitly consider it as WAIT ALL? (Or to be honest I think it could also return an error, since it will obviously just timeout, i don't think that waiting for a replica that will be online in the future is a normal user request.)

I have encountered such a problem before.
Redisson sent a `wait numreplicas timeout` command, where the value of numreplicas was much larger than the actual number of replicas, causing the client to be blocked and timeout (and return an error).

There are several reasons for the value being much larger.
1. There are problems with the client code, such as https://github.com/redisson/redisson/issues/1304, which counts multiple replicas.
2. Or the cluster slots returned in the proxy implementation are not standard.

I remember we had such a problem with sentinel. The failover it triggered would not pause the primary, causing the new replica to lose some data during the failover. I want to deal with this before, but i haven't found the time yet.

**Comment by ranshid** [comment]
I like this proposal very much as it greatly simplifies existing scenarios for clients.

I think, though, that it should be made very clear what is the guarantee of this new API. For example what happen in case of syncing replica? A replica which is waiting for BGSAVE or consuming the RDB will not receive a getack request from the primary so it should not report back to the server thus wait will timeout right? Maybe it should be made clear that we will only wait for the "ACTIVE" replicas which are in the process of steady state syncing or psync?   

ps. I also noticed that there was some proposal to have a strict configurable value of "num or replicas to sync". We should be able to handle different scenarios of upgrade cluster, where sometimes additional replicas are added in order to later failover to.

**Comment by enjoy-binbin** [comment]
> Maybe it should be made clear that we will only wait for the "ACTIVE" replicas which are in the process of steady state syncing or psync?

Yeah, i think we should aim for "ACTIVE" replicas. But it is also hard to say, if the replica happens to be disconnected due to the network, does WAIt need to wait unit it comes online?

> We should be able to handle different scenarios of upgrade cluster, where sometimes additional replicas are added in order to later failover to.

So when we removing a replica, change the "value" of the majority / all? or we can dynamically calculate the current majority / all value each time instead of relying on the value calculated at the very first WAIT command.

**Comment by zuiderkwast** [comment]
> if the replica happens to be disconnected due to the network, does WAIt need to wait unit it comes online?

Yes it does. Otherwise it's just best effort. What I'm interested in is a real guarantee. It would make valkey cluster a database with real concistency. Then, you can store important data and not just cache and session data.

**Comment by zuiderkwast** [comment]
When counting replicas in cluster mode, we need to count all replicas that can win a failover election. For example, we need to count also temporarily disconnected replicas, based on cluster node information.

If a relica is new and has the offset 0 it will not start a failover election? In that case maybe we can exclude it. I don't know this logic exactly. The point is that we can ignore it only if it's not eligible for failover.

**Comment by jamesx-improving** [comment]
Hi @enjoy-binbin @ranshid , any plan to implement this? Valkey-GLIDE has a client side [user issue](https://github.com/valkey-io/valkey-glide/issues/3329) that depends on this.

**Comment by ranshid** [comment]
> Hi [@enjoy-binbin](https://github.com/enjoy-binbin) [@ranshid](https://github.com/ranshid) , any plan to implement this? Valkey-GLIDE has a client side [user issue](https://github.com/valkey-io/valkey-glide/issues/3329) that depends on this.

@jamesx-improving thank you for the wakeup call. I suggest we need to think about it for Valkey-10 in the scope of the planned: https://github.com/valkey-io/valkey/issues/3037

I mean sync-replicaiton might provide this functionality right?

**Comment by zuiderkwast** [comment]
Yeah, WAIT ALL is basically opt-in sync replication. The question is whether WAIT ALL can provide durability and strong consistency guarantees.

If all replicas disconnect, ALL means zero, so WAIT ALL will succeed immediately if there are no replicas. That's not very useful.

If it's combined with the config `min-replicas-to-write 1`, the server will refuse to write if no replicas are connected. Better. But is it enough in all cases?

What happens during rolling upgrade? Let's say, in normal operation, there is one primary P and one replica R1. During a rolling upgrade, a new replica R2 is added. Then, the old R1 is removed. If, just before R1 is removed, R2 is disconnected or crashes. Then, some writes only reach R1, and WAIT ALL succeeds because there is only one replica connected, but then R1 is shutdown and removed. Some writes are not replicated, even though the writes and WAIT ALL have succeeded. Not what we wanted.

What if the administrator sets `min-replicas-to-write 2` during the upgrade? If it's done just before adding R2 and set back to 1 just after removing R1, then this seems safe to me!

In cluster mode, it would also be possible to count WAIT ALL based on the cluster replicas, including the ones that are disconnected and marked as failing, until you call CLUSTER FORGET. This would be similar to adjusting `min-replicas-to-write` during the upgrade.

Could this be a minimal step towards sync replication and strong consistency?

**Comment by murphyjacob4** [comment]
> it would also be possible to count WAIT ALL based on the cluster replicas, including the ones that are disconnected and marked as failing

Another thing to consider is the number of "cluster replicas" can be different than the number of "active PSYNCs". E.g. running tools like RedisShake or nodes that are forgotten from the cluster topology but still replicating. Just counting REPLCONF ACKs isn't enough, we need to ensure we match those REPLCONF ACKs to a certain membership set. 

Dynamic membership is another class of problem you mention. Who is considered part of "ALL" may be changing. For strong consistency via synchronous replication, it must bet true that when `WAIT ALL` returns successfully, all potential future primaries of the cluster have recorded the write. But `WAIT ALL` does not guarantee that, it only guarantees the current set of replicas have recorded the write. A new replica could be added after the `WAIT ALL` returns, and it could successfully `PSYNC` without having recorded the write, and thus still be eligible for primaryship.
